[Project] Up Next Shuffle - Implement Shuffle Action

mebarbosa opened the pull request:
## Description
- This PR add the action to shuffle when tapping on up next shuffle button

> [!NOTE]  
> In another PR I will handle the shuffle state sync

Fixes #3064

## Testing Instructions
Have the shuffle feature flag enabled

### Shuffle Feature

1. Add some episodes to Up Next
2. Play an episode
3. Have the shuffle disabled
4. Seek the current playing episode to almost the end
5. Wait for the episode ends
6. ✅ Ensure the first episode from up next starts playing
7. Enable shuffle
8. Seek the current playing episode to almost the end
9. Wait for the episode ends
10. ✅ Ensure a random episode from up next starts playing _(Since we are using a random way to get the next episode, we can also pick the first one from up next, so I recommend to test this a couple of times to ensure that is taking random episodes from up next)_
11. ✅ Ensure the random episode that started playing was removed from the up next queue

### Regression tests

#### Play Next / Play Last
1. Have some episodes in Up next and shuffle enabled
2. Open an podcast
3. Swipe to add episode to Up next
4. ✅ Ensure you can add episodes to both end of the queue and top of the queue

<img width="356" alt="image" src="https://github.com/user-attachments/assets/fc243577-62b8-4b96-8091-5effb7b75ac0">

#### Play Up Next episode on Tap
1. Have some episodes in Up next and shuffle enabled
2. Go to Profile -> Settings -> General -> Enable `Play Up Next episode on Tap`
3. Go to Up next
4. Tap on some episode from Up next on queue
5. ✅ Ensure the episode you tapped starts playing
6. ✅ Ensure the up next kept the episode sort (only changed the episode you tapped and the current playing episode)

#### Sync Up next Queue
1. Have two devices logged with the same account
2. Have some populated Up Next
3. Tap on Refresh now in both of them to ensure they are synced
4. Ensure the two devices have the same up next queue
5. On device A, enable shuffle, wait for the current episode finishes
6. ✅ Ensure a random episode was picked
7. Observe the current Up Next
8. Go to Profile and Refresh Now
9. On device B, go to Profile and Refresh Now
10. Check the Up Next
11. ✅ Ensure the Up Next is the same you checked on step 7




## Screenshots or Screencast 
[Screen_recording_20241022_161358.webm](https://github.com/user-attachments/assets/5e93b8af-c9f0-4cf9-a6d3-ed2c792ed091)


## Checklist
- [ ] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [ ] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [ ] I have considered whether it makes sense to add tests for my changes
- [ ] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [ ] Any jetpack compose components I added or changed are covered by compose previews
- [ ] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.
 
#### I have tested any UI changes...
<!-- If this PR does not contain UI changes, ignore these items -->
- [ ] with different themes
- [ ] with a landscape orientation
- [ ] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack

dangermattic commented:
<!--
  1 Error: Please add tests for class <co...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>RemoveAndShuffle</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

MiSikora commented on modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/UpNextQueueImpl.kt:
```diff
@@ -232,9 +238,13 @@ class UpNextQueueImpl @Inject constructor(
     }
 
-    override suspend fun removeEpisode(episode: BaseEpisode) {
+    override suspend fun removeEpisode(episode: BaseEpisode, shouldShuffleUpNext: Boolean) {
```
There's no need to pass `shouldShuffleUpNext` to this function. You already have access to settings that hold the property you can read. This also simplifies a lot of things because there will be no need to for `RemoveAndShuffle` type since everything can be determined by reading the user setting.

MiSikora reviewed (changes requested):
Overall works well but the code can be simplified by not passing the `shuffle` state between different layers.

There's also an issue when an episode is archived or marked as played while listening to it. If that happens the next episode isn't chosen at random.

geekygecko commented on modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/UpNextAdapter.kt:
```diff
@@ -173,30 +173,43 @@ class UpNextAdapter(
-        }
+            val isEnabled = settings.upNextShuffle.value
+
+            this.setImageDrawable(
+                ContextCompat.getDrawable(
+                    context,
+                    if (isEnabled) IR.drawable.shuffle_enabled else IR.drawable.shuffle,
+                ),
+            )
```
```suggestion
            this.setImageResource(if (isEnabled) IR.drawable.shuffle_enabled else IR.drawable.shuffle)
```

Can you use `setImageResource` instead here?

mebarbosa commented:
Hey @MiSikora . Thank you for testing and catching the played / archived scenario. I've addressed them already and updated the PR description with more scenarios to test. When you have time, could you please test them again?

MiSikora reviewed (approved):
I found one minor thing but I'm not sure if it is even an issue. When I archive or mark an episode not from the player but, for example, from a podcast's episode list the Up Next doesn't use shuffle. But I think it is a very obscure scenario that we might not even want to support.

mebarbosa commented:
> I found one minor thing but I'm not sure if it is even an issue. When I archive or mark an episode not from the player but, for example, from a podcast's episode list the Up Next doesn't use shuffle. But I think it is a very obscure scenario that we might not even want to support.

I will check this scenario as well and address in another PR if we decided to handle. Thank you again!
